### PR TITLE
[HTTP2_REPLICATION] Changes to support http2 Replication

### PR DESCRIPTION
**Pull request by zzmao**
Use `clustermap.enable.http2.replication` as config to enable/disable http2 based replication.

### CODE REVIEW

File: ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java
Comment (justinlin-linkedin) [inline ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java:729]: this not going to work, since the underlying input stream for getResponse is a `DataInputStream` not a `NettyByteBufDataInputStream`. check out line 1072.

We can make ChannelOutput return `DataInputStream` in `getInputStream` method, this would fix the problem.

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java:101]: logger? or maybe remove?

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java:106]: same above.

File: ambry-network/src/main/java/com/github/ambry/network/http2/Http2BlockingChannel.java
Comment (cgtz) [inline ambry-network/src/main/java/com/github/ambry/network/http2/Http2BlockingChannel.java:117]: replace 8 with Long.BYTES

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServer.java
Comment (cgtz) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryServer.java:231]: Do you think we will eventually use DataNodeId.getPortToConnectTo in the router too? If so, we may want to change the config name to `clusterMapEnableHttp2Client` (and deprecate the router config at some point) so that it makes sense for both the router and replication.

Comment (zzmao) [reply ambry-server/src/main/java/com/github/ambry/server/AmbryServer.java:231]: Yes. 

I had an old comment:
```
  /**
   * A temporary method to get port before http2 based replication is implemented.
   * TODO: remove this once http2 based replication is implemented.
   * @param replicaId The {@link ReplicaId} to connect to.
   * @param routerEnableHttp2NetworkClient  if http2 network client should be enabled.
   * @return the port to connect.
   */
  static Port getPortToConnectTo(ReplicaId replicaId, boolean routerEnableHttp2NetworkClient) {
    if (routerEnableHttp2NetworkClient) {
      return new Port(replicaId.getDataNodeId().getHttp2Port(), PortType.HTTP2);
    } else {
      return replicaId.getDataNodeId().getPortToConnectTo();
    }
  }
```

The plan is:
1. Apply this PR and verify http2 replication works
2. Deprecate `sslEnabledDataCenters` and `routerEnableHttp2NetworkClient`. Rename `clusterMapEnableHttp2Replication` to `clusterMapEnableHttp2`. 

After deprecating `sslEnabledDataCenters`, the default connection is SSL based, if clusterMapEnableHttp2 is enabled, it's Http2 based. 

Both frontend and server will be accessible to  `clusterMapEnableHttp2`.

**Review by justinlin-linkedin — APPROVED** [review]

**Comment by zzmao** [comment]
```
2020-12-09 11:59:31,940] ERROR LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
Created at:
	io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:385)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:187)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:178)
	io.netty.handler.ssl.SslHandler.allocate(SslHandler.java:2196)
	io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1375)
	io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1275)
	io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1322)
	io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501)
	io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:440)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.lang.Thread.run(Thread.java:745) (io.netty.util.ResourceLeakDetector)
[2020-12-09 11:59:31,944] ERROR LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
Created at:
	io.netty.buffer.AbstractByteBufAllocator.compositeDirectBuffer(AbstractByteBufAllocator.java:223)
	io.netty.buffer.AbstractByteBufAllocator.compositeBuffer(AbstractByteBufAllocator.java:201)
	io.netty.handler.codec.MessageAggregator.decode(MessageAggregator.java:271)
	io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	io.netty.handler.codec.http2.AbstractHttp2StreamChannel$Http2ChannelUnsafe.doRead0(AbstractHttp2StreamChannel.java:901)
	io.netty.handler.codec.http2.AbstractHttp2StreamChannel.fireChildRead(AbstractHttp2StreamChannel.java:555)
	io.netty.handler.codec.http2.Http2MultiplexHandler.channelRead(Http2MultiplexHandler.java:179)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	io.netty.handler.codec.http2.Http2FrameCodec.onHttp2Frame(Http2FrameCodec.java:652)
	io.netty.handler.codec.http2.Http2FrameCodec$FrameListener.onHeadersRead(Http2FrameCodec.java:594)
	io.netty.handler.codec.http2.Http2FrameCodec$FrameListener.onHeadersRead(Http2FrameCodec.java:588)
	io.netty.handler.codec.http2.Http2FrameListenerDecorator.onHeadersRead(Http2FrameListenerDecorator.java:48)
	io.netty.handler.codec.http2.Http2EmptyDataFrameListener.onHeadersRead(Http2EmptyDataFrameListener.java:63)
	io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:373)
	io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:321)
	io.netty.handler.codec.http2.Http2InboundFrameLogger$1.onHeadersRead(Http2InboundFrameLogger.java:56)
	io.netty.handler.codec.http2.DefaultHttp2FrameReader$2.processFragment(DefaultHttp2FrameReader.java:483)
	io.netty.handler.codec.http2.DefaultHttp2FrameReader.readHeadersFrame(DefaultHttp2FrameReader.java:491)
	io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:254)
	io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:160)
	io.netty.handler.codec.http2.Http2InboundFrameLogger.readFrame(Http2InboundFrameLogger.java:41)
	io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:174)
	io.netty.handler.codec.http2.DecoratingHttp2ConnectionDecoder.decodeFrame(DecoratingHttp2ConnectionDecoder.java:63)
	io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:378)
	io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:438)
	io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501)
	io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:440)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1526)
	io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1275)
	io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1322)
	io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501)
	io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:440)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.lang.Thread.run(Thread.java:745) (io.netty.util.ResourceLeakDetector)
```

A leak found, please don't merge it for now.

**Comment by zzmao** [comment]
@cgtz @jsjtzyy  After debugging the leak with @justinlin-linkedin , we ruled out prod code memory leak. The leak is from test, but we can't locate it, because the access is from sslHandler, we have no more clue to debug. 

I am thinking if we can merge this PR and verify it in EI. The test leak is noted in test comment.

**Comment by zzmao** [comment]
@cgtz @justinlin-linkedin @jsjtzyy  The memory leak is final fixed. There were lots of leak from unit test that caused test stuck.
The most recent changes are memory leak fix in unit test by releasing the bytebuf in the stream.

Ready for final review.

File: ambry-api/src/main/java/com/github/ambry/network/ChannelOutput.java
Comment (jsjtzyy) [inline ambry-api/src/main/java/com/github/ambry/network/ChannelOutput.java:41]: remove this?

Comment (zzmao) [reply ambry-api/src/main/java/com/github/ambry/network/ChannelOutput.java]: my latest changes were not in. just pushed.

**Review by justinlin-linkedin — APPROVED** [review]

File: ambry-server/src/integration-test/java/com/github/ambry/server/CloudAndStoreReplicationTest.java
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com/github/ambry/server/CloudAndStoreReplicationTest.java:304]: nit: remove necessary `DataInputStream` import.

File: ambry-server/src/integration-test/java/com/github/ambry/server/DirectSender.java
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com/github/ambry/server/DirectSender.java:29]: same here for `InputStream`

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerHardDeleteTest.java
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerHardDeleteTest.java:393]: Minor: can be simplified to
```
PutResponse response0 = PutResponse.readFrom(channel.sendAndReceive(putRequest0).getInputStream());
```

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerHardDeleteTest.java
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerHardDeleteTest.java:407]: same here

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerHardDeleteTest.java
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerHardDeleteTest.java:421]: same here

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerHardDeleteTest.java
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerHardDeleteTest.java:460]: same here (I noticed other files may be simplified too, it's ok to defer such simplification to future PR)

**Review by jsjtzyy — APPROVED** [review]
LGTM after a few minor comments are addressed

File: ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java:2869]: nit: copy paste error?

**Review by jsjtzyy — APPROVED** [review]

**Pull request merged**
